### PR TITLE
Training & fine-tuning quickstart

### PR DESCRIPTION
**Pull request by joeddav**
This PR adds a short guide to the docs demonstrating how to train & fine-tune models in native PyTorch, native TF2, and using Trainer.

My aim was not to show how to train on every type of task, but simply to communicate the key points along with a couple of very simple and easy-to-follow examples. More involved examples spanning many tasks are linked at the bottom.

### CODE REVIEW

File: docs/source/training.rst [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
+    ]
+    optimizer = AdamW(optimizer_grouped_parameters, lr=1e-5)
+    scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps, num_train_steps)
+
+    # iterate over batches of preprocessed training data
+    for batch in train_loader:
+        optimizer.zero_grad()
+        
+        input_ids = batch.input_ids.to(device)
```
Comment (sshleifer) [inline docs/source/training.rst:105]: you can call `batch.to(device)`

**Review by sshleifer — APPROVED** [review]
LGTM

File: docs/source/training.rst [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    from transformers import BertTokenizer, BertForSequenceClassification, AdamW
+
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
+    model = BertForSequenceClassification.from_pretrained('bert-base-uncased')
+    model.train()
+    optimizer = AdamW(model.parameters(), lr=5e-5) # any PT optimizer works
+
+    text_batch = ["I love Pixar.", "I don't care for Pixar."]
+    labels = torch.tensor([1,0]).unsqueeze(0)
+    encoding = tokenizer.batch_encode_plus(text_batch, return_tensors='pt',
```
Comment (thomwolf) [inline docs/source/training.rst:67]: Can you switch to the new API?

**Review by thomwolf — APPROVED** [review]
Nice!

File: docs/source/training.rst [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
+    loss = outputs[0]
+    loss.backward()
+    optimizer.step()
+
+Of course, you can train on GPU by calling ``to('cuda')`` on the model and
+inputs as usual.
+
+We also provide a few tools to make it easier to apply more complex
+fine-tuning procedures. Here's an example using weight decay and learning
```
Comment (sgugger) [inline docs/source/training.rst:81]: "using weight decay (except for the bias or layer norm parameters) and"

File: docs/source/training.rst (hunk 1 above)
Comment (sgugger) [reply docs/source/training.rst:105]: This is where the "we omit the data preparation part" is annoying. Previously the batch was a dict, now it's something that has an `input_ids` attribute and even a magic method.

If the data is prepared the same way as before, it should call `input_ids = encoding['input_ids']` like before. If it's not, the preparation step should be presented separately (and maybe explain why there is that magic `to` method.

File: docs/source/training.rst [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    optimizer = AdamW(model.parameters(), lr=5e-5) # any PT optimizer works
+
+    text_batch = ["I love Pixar.", "I don't care for Pixar."]
+    labels = torch.tensor([1,0]).unsqueeze(0)
+    encoding = tokenizer.batch_encode_plus(text_batch, return_tensors='pt',
+                                           pad_to_max_length=True)
+    input_ids = encoding['input_ids']
+    attention_mask = encoding['attention_mask']
+
+    outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
```
Comment (sgugger) [inline docs/source/training.rst:72]: It is mentioned before but I would put a comment here with the fact that we pass the `labels` to get back the `loss` directly. This might surprise someone who is not familiar with the library yet.

Or to avoid comments, maybe we could split the codes in several smaller pieces: for now this is all the explanations then all the code, so the reader forgets part of the explanation when they reach the middle of the code. Maybe we could have use a more progressive approach and  split the code in smaller bits with their explanation just before?

File: docs/source/training.rst [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+
+.. code-block:: python
+
+    from transformers import get_linear_schedule_with_warmup
+
+    device = torch.device('cuda')
+    model = BertForSequenceClassification.from_pretrained('bert-base-uncased').to(device)
+    model.train()
+
+    # Prepare optimizer and schedule (linear warmup and decay)
```
Comment (sgugger) [inline docs/source/training.rst:92]: Same as before, consider splitting the code at each commented line and introduce in prose what happens after.

File: docs/source/training.rst [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+mixed precision and easy tensorboard logging.
+
+
+Freezing the encoder
+--------------------
+
+In some cases, you might be interested in keeping the weights of the
+pre-trained encoder frozen and optimizing only the weights of the head
+layers. To do so, simply set the ``requires_grad`` attribute to ``False`` on
+the encoder parameters, which can be accessed with the ``base_model``
```
Comment (sgugger) [inline docs/source/training.rst:126]: I would add a remark that `base_model` works for every task-specific model in the library. It's a useful convention for the reader to know.

File: docs/source/training.rst [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+Keras ``fit`` method. Here's an example of fine-tuning a sequence classifier
+with a pre-trained encoder on MRPC:
+
+.. code-block:: python
+    
+    import tensorflow as tf
+    import tensorflow_datasets
+    from transformers import TFBertForSequenceClassification, BertTokenizer, glue_convert_examples_to_features
+
+    # Load dataset, tokenizer, model from pretrained model/vocabulary
```
Comment (sgugger) [inline docs/source/training.rst:153]: Same comment as before.

File: docs/source/training.rst [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+    model.compile(optimizer=optimizer, loss=loss)
+
+    # Train and evaluate using tf.keras.Model.fit()
+    history = model.fit(train_dataset, epochs=2, steps_per_epoch=115,
+                        validation_data=valid_dataset, validation_steps=7)
+
+This example uses the built-in
+:func:`~transformers.data.processors.glue.glue_convert_examples_to_features`
+to tokenize MRPC and convert it to a TF ``Dataset`` object which can be
```
Comment (sgugger) [inline docs/source/training.rst:175]: Tf -> TensorFlow (in prose, let's not abbreviate it :-) )

File: docs/source/training.rst [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+passed to ``model.fit()``.
+
+With the tight interoperability between TensorFlow and PyTorch models, you
+can even save the model and then reload it as a PyTorch model:
```
Comment (sgugger) [inline docs/source/training.rst:179]: This is very cool. Does it work backward too? (e.g. can I load a PyTorch-pretrained model back in TF?) Mention the answer either way.

I think this is a cool feature so it should be mentioned in the intro, the reader might not know of it.

File: docs/source/training.rst [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+We also provide a simple but feature-complete training and evaluation
+interface through :func:`~transformers.Trainer` and
+:func:`~transformers.TFTrainer`. You can train, fine-tune,
+and evaluate any Transformers model with a wide range of training options and
+with built-in features like logging, gradient accumulation, and mixed
+precision.
+
+Using the trainer requires first creating a model and defining a
+:func:`~transformers.data.DataCollator`, which is responsible for taking in a batch and preparing
```
Comment (sgugger) [inline docs/source/training.rst:200]: It's actually become a simple function now, so you should use it to simplify the example.

File: docs/source/training.rst [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+call our model. Below, we define a collator in the case where a batch
+consists of a list of dicts and returns a single dict with the examples
+concatenated.
+
+.. code-block:: python
+
+    from transformers import DataCollator
+
+    class MyCollator(DataCollator):
+        def collate_batch(self, batch):
```
Comment (sgugger) [inline docs/source/training.rst:212]: This can become just a function `my_data_collator`, also mention that `default_data_collator` would do this too (or at least it will very soon ^^)

File: docs/source/training.rst [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+            attention_mask = torch.stack([example['attention_mask'] for example in batch])
+            labels = torch.stack([example['label'] for example in batch])
+
+            return {
+                'input_ids': input_ids,
+                'attention_mask': attention_mask,
+                'labels': labels
+            }
+
+Now, you can simply define a :func:`~transformers.Trainer` and
```
Comment (sgugger) [inline docs/source/training.rst:223]: This is exactly what I have in mind when I suggested splitting the code on commented lines before, it works much better in a style like this for a tutorial.

File: docs/source/training.rst [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+:func:`~transformers.TrainingArguments` and the all
+of the underlying work is done for you.
+
+.. code-block:: python
```
Comment (sgugger) [inline docs/source/training.rst:227]: Could we use the same style as the usage page and have this block able to switch between PyTorch and TensorFlow? That would be super cool.

**Review by sgugger — CHANGES_REQUESTED** [review]
This is great work! 
I left a few suggestions, we can address some of them in subsequent PRs if that's easier. The only thing to change right now would then be to use the functional style for the data collator.

File: docs/source/training.rst (hunk 10 above)
Comment (joeddav) [reply docs/source/training.rst:222]: It looks like `Trainer` still expects an object with a `collate_batch` method though. Thoughts @julien-c?

File: docs/source/training.rst (hunk 11 above)
Comment (joeddav) [reply docs/source/training.rst]: I opted to just remove the data collating since most cases shouldn't need a custom one now. Lmk what you think.

**Comment by joeddav** [comment]
@patrickvonplaten This is a little bit less verbose than what I think you were envisioning but curious what you think. I just showed sequence classification to communicate the general principles rather than covering multiple different tasks, which would make things pretty long.

File: docs/source/training.rst [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+
+Model classes in 🤗 Transformers are designed to be compatible with native
+PyTorch and TensorFlow 2 and can be used seemlessly with either. In this
+quickstart, we will show how to fine-tune (or train from scratch) a model
+using the standard training tools available in either framework. We will also
+show how to use our included :func:`~transformers.Trainer` class which
+handles much of the complexity of training for you.
+
+This guide assume that you are already familiar with loading and use our
+models for inference; otherwise, see `Usage <./usage.html>`_. We also assume
```
Comment (sgugger) [inline docs/source/training.rst:12]:
```suggestion
models for inference; otherwise, see the :doc:`task summary <task_summary>`. We also assume
```

**Review by sgugger — APPROVED** [review]
Just one link to change since the doc has been reorganized a bit.

**Comment by patrickvonplaten** [comment]
I think it's great! At the moment we can't really show causal lm training because it's not implemented yet in TF :D So sequence classification sounds good to me! 

In the longer term, think we should have one section for each model type for both TF and PT:

- Causal LM

- Masked LM

- Seq2Seq

- Seq classifaction

....

But for now I like it!

**Comment by patrickvonplaten** [comment]
I don't think, the section would become too long if we show training for every model type (CLM, MLM, Seq2Seq, ...). If we would start with CLM and only add a couple of sentences for what is different from the previous section with some code could be very helpful!

I'm still wondering if we should have a training section on each model page since training can differ quite a lot between models: XLNet has a very special training scheme, T5 pretraining is different from Bart, ...

**Review by patrickvonplaten — APPROVED** [review]

**Comment by sgugger** [comment]
I personally think this is okay if each task is shown in a different notebook/tutorial (there is a big table of tasks after all). When we are at a point where all tasks can be easily loaded in a few lines of code we can maybe show more, but I fear that the specificity of each task/dataset requiring its own preprocessing function will lose the reader when the essential point of this (beginner's) tutorial is on training and Trainer/TFTrainer.

Having an example on each model page may also be problematic since models can be used for several tasks. So it might turn up in having way too many things in the docs as well. For now I think making more independent notebooks that show how to train/fine-tune a model on a given task and link to those in all the right places might be the best solution. That way the reader opts in to see this model trained on that task.

**Comment by joeddav** [comment]
I think it's a legitimate question how much guidance we should give for more obscure cases that you mentioned, @patrickvonplaten. My feeling is that those are fairly specialized and it's fair to expect users to be able to figure out more advanced cases like that out between docs/source code/table of tasks. I wouldn't be opposed to incorporating some of the more common tasks (e.g. MLM training), but I generally agree with @sgugger to err on the side of brevity and clarity for the purpose of a quickstart guide like this one. Then we can lean on a combination of model docs and the big table of tasks for the more obscure/specialized cases.

Also, it would help to have docs for Trainer.
